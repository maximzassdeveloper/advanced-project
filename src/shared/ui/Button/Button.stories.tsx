import { Meta } from '@storybook/react'
import { Button } from './Button'

const meta: Meta<typeof Button> = {
  title: 'shared/Button',
  component: Button,
}
export default meta

export const Themes = () => (
  <div>
    <Button theme='primary'>Primary</Button>
    <Button theme='outline'>Outline</Button>
    <Button theme='clear'>Clear</Button>
  </div>
)

export const Sizes = () => (
  <div>
    <Button size='m'>Button M</Button>
    <Button size='s'>Button S</Button>
    <Button size='xs'>Button XS</Button>
  </div>
)

export const Disabled = () => (
  <div>
    <Button disabled theme='primary'>
      Primary
    </Button>
    <Button disabled theme='outline'>
      Outline
    </Button>
    <Button disabled theme='clear'>
      Clear
    </Button>
  </div>
)

export const Loading = () => (
  <div>
    <Button loading>Disabled</Button>
  </div>
)

export const Focused = () => (
  <div>
    <Button autoFocus>Focused</Button>
  </div>
)
