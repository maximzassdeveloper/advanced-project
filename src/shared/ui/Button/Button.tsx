import { ButtonHTMLAttributes, forwardRef, useEffect, useRef } from 'react'
import { CSSTransition } from 'react-transition-group'
import { Spinner } from '@/shared/ui'
import { useAddWave } from '@/shared/hooks'
import { composeRef } from '@/shared/lib/composeRef'
import { classNames } from '@/shared/lib/classNames'
import { mergeProps } from '@/shared/lib/mergeProps'
import s from './button.module.scss'

type ButtonTheme = 'primary' | 'outline' | 'clear'
type ButtonSize = 'xs' | 's' | 'm'

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  theme?: ButtonTheme
  size?: ButtonSize
  loading?: boolean
  className?: string
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    children,
    type = 'button',
    className,
    size = 's',
    theme = 'primary',
    loading,
    disabled,
    ...rest
  } = props

  const buttonRef = useRef<HTMLButtonElement>(null)
  const waveProps = useAddWave<HTMLButtonElement>(buttonRef, props)

  return (
    <button
      ref={composeRef(buttonRef, ref)}
      type={type}
      className={classNames(s.button, className, s[size], s[theme], { [s.loading]: loading })}
      disabled={loading || disabled}
      {...mergeProps(rest, waveProps)}
    >
      {children}
      <CSSTransition timeout={200} in={loading} classNames='fade' mountOnEnter unmountOnExit>
        <span className={s.loader}>
          <Spinner size='small' color='white' />
        </span>
      </CSSTransition>
    </button>
  )
})
