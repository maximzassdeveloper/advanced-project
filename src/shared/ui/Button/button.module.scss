.button {
  position: relative;
  display: inline-block;
  white-space: nowrap;
  touch-action: manipulation;
  cursor: pointer;
  user-select: none;
  border-radius: 3px;
  border: 1px solid transparent;
  transition: all 0.2s;

  &:disabled {
    pointer-events: none;
    opacity: 0.8;
  }

  @include focus-visible-outline;

  &.m {
    font: var(--font-m);
    font-weight: 700;
    padding: 9px 30px;
  }

  &.s {
    font: var(--font-s);
    font-weight: 700;
    padding: 8px 30px;
  }

  &.xs {
    font: var(--font-xs);
    font-weight: 500;
    padding: 6px 14px;
  }

  &.primary {
    color: var(--primary-color);
    background: var(--primary);

    &:hover:not(:disabled) {
      background: var(--primary-hover);
    }
  }

  &.outline {
    color: var(--primary);
    border: 1px solid var(--primary);
    background: transparent;
    font-weight: 500;

    &:hover:not(:disabled) {
      border-color: var(--primary-hover);
      color: var(--primary-hover);
    }
  }

  &.clear {
    border: none;
    background: transparent;
    font-weight: 500;

    &:hover:not(:disabled) {
      text-decoration: underline;
    }
  }

  &.icon {
    height: 45px;
    min-width: 45px;
    width: 45px;
    padding: 4px !important;
    display: flex;
    align-items: center;
    justify-content: center;

    i {
      color: var(--secondary);
      font-size: 30px;
    }
  }
}

.loader {
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  background: inherit;
  border-radius: inherit;
}
