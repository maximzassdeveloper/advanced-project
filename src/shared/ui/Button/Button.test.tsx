import { useState } from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { Button, ButtonProps } from './Button'

const WrapperToTestLoading = (props: ButtonProps) => {
  const { loading, disabled, ...rest } = props
  const [_disabled, setDisabled] = useState(disabled)
  const [_loading, setLoading] = useState(loading)

  return (
    <>
      <span onClick={() => setLoading((p) => !p)}>Trigger loading</span>
      <span onClick={() => setDisabled((p) => !p)}>Trigger disabled</span>
      <Button disabled={_disabled} loading={_loading} {...rest}>
        Button
      </Button>
    </>
  )
}

describe('Button', () => {
  test('should return correct snapshot', () => {
    const { container } = render(<Button>Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('render correctly primary theme', () => {
    const { container } = render(<Button theme='primary'>Button</Button>)
    expect(container.firstChild).toHaveClass('primary')
  })

  test('render correctly disabled', () => {
    const { container } = render(<Button disabled>Button</Button>)
    expect(container.firstChild).toBeDisabled()
  })

  test('should not clickable when button is loading', () => {
    const onClick = jest.fn()
    render(
      <Button loading onClick={onClick}>
        Button
      </Button>
    )
    const button = screen.getByRole('button')

    fireEvent.click(button)
    expect(onClick).toHaveBeenCalledTimes(0)
  })

  test('should make actual disable state after loading', () => {
    const onClick = jest.fn()

    render(<WrapperToTestLoading disabled />)

    const button = screen.getByRole('button')
    const triggerLoading = screen.getByText(/Trigger loading/i)

    fireEvent.click(triggerLoading)
    fireEvent.click(button)
    expect(onClick).toHaveBeenCalledTimes(0)

    fireEvent.click(triggerLoading)
    fireEvent.click(button)
    expect(onClick).toHaveBeenCalledTimes(0)
    expect(button).toBeDisabled()
  })

  test('should make actual state if disabled is changed during loading', () => {
    const onClick = jest.fn()

    render(<WrapperToTestLoading disabled />)

    const button = screen.getByRole('button')
    const triggerLoading = screen.getByText(/Trigger loading/i)
    const triggerDisable = screen.getByText(/Trigger disable/i)

    // disabled and loading
    fireEvent.click(triggerLoading)
    fireEvent.click(button)
    expect(onClick).toHaveBeenCalledTimes(0)

    // not disabled and loading
    fireEvent.click(triggerDisable)
    expect(button).toBeDisabled()

    // not disabled and not loading
    fireEvent.click(triggerLoading)
    fireEvent.click(button)
    expect(button).not.toBeDisabled()
  })
})

