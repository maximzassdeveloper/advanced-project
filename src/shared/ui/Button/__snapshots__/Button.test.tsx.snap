// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should return correct snapshot 1`] = `
<button
  class="button s primary"
  type="button"
>
  Button
</button>
`;
